Fixed app session start audit events for Device Trust failures

21KennethTran opened the pull request:
When Device Trust and MFA is enabled for accessing a resource and a user without a trusted device tries to access that resource, their session should get rejected and an auth attempt fail event is generated in the audit log.

Since this type of session rejection is an auth failure, the type of event generated in the audit logs should be an AppSessionStart. The AppMetadata struct is added to this to provide more information about the resource the user tried to access.

Following the recreation steps from #54139 ,

Before:

https://github.com/user-attachments/assets/7e36fc68-ca2e-41d3-a61f-ca4a53b9651c


After:

https://github.com/user-attachments/assets/e9610592-69fe-42ad-8d89-1bc5563364a3

Here is the audit log fields with the changes from the demo above:
```
{
  "app_name": "debug-app",
  "app_public_addr": "debug-app.localhost",
  "app_uri": "http://localhost:8000/",
  "cluster_name": "localhost",
  "code": "T2007E",
  "ei": 0,
  "error": "access to resource requires a trusted device",
  "event": "app.session.start",
  "message": "access to application debug-app requires a trusted device",
  "namespace": "default",
  "public_addr": "",
  "server_id": "4948fc26-21e7-4aaf-a45b-02b39d392a0d",
  "server_version": "19.0.0-dev",
  "sid": "",
  "time": "2026-01-21T00:27:54.033Z",
  "uid": "b53c7358-bdc9-4a99-94a7-0a56477fa2f1",
  "user": "kenneth.tran@goteleport.com",
  "user_cluster_name": "localhost",
  "user_kind": 1,
  "user_roles": [
    "access",
    "editor",
    "mfa-device-trust",
    "require-trusted-device"
  ],
  "user_traits": {
    "acr": [
      "1"
    ],
    "at_hash": [
      "_J7AxLN1qCDTGoeMQp5y8Q"
    ],
    "aud": [
      "teleport-oidc"
    ],
    "azp": [
      "teleport-oidc"
    ],
    "email": [
      "kenneth.tran@goteleport.com"
    ],
    "family_name": [
      "tran"
    ],
    "given_name": [
      "kenneth"
    ],
    "groups": [
      "teleport-admins"
    ],
    "iss": [
      "https://localhost:8443/realms/teleport"
    ],
    "jti": [
      "38610d7d-fad0-c7de-03c2-60b1b4cacd4c"
    ],
    "name": [
      "kenneth tran"
    ],
    "preferred_username": [
      "kenneth-keycloak"
    ],
    "sid": [
      "3c62224c-e0bf-37b9-5c57-a76a2dd53550"
    ],
    "sub": [
      "80814091-50d6-4214-8b23-229a1f13c596"
    ],
    "typ": [
      "ID"
    ],
    "username": [
      "kenneth-keycloak"
    ]
  },
  "with_mfa": "77879faa-6ed8-4799-8b0e-e46af9f3facb"
}
```


changelog: Reject application session requests and emit a failed session start audit event when a trusted device is required but not provided

21KennethTran committed:
Created check for device trust after MFA before creating web session

zmb3 commented on api/proto/teleport/legacy/types/events/events.proto:
I wonder if we should change this doc comment.

Are there any cases where this is being emitted on success?

21KennethTran replied:
There doesn't seem to be any instances where an `AuthAttempt` has a success in the code field. It seems that successful events have more granular types and success codes. I will change the doc comment for more clarity.

zmb3 commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -2187,6 +2165,13 @@ message AuthAttempt {
     (gogoproto.jsontag) = ""
   ];
+
+  // App is a common application resource metadata.
```
```suggestion
  // App hold information about the target application.
```

zmb3 commented on lib/auth/sessions.go:
```diff
@@ -544,6 +539,55 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
+					ServerVersion:   teleport.Version,
+					ServerID:        a.ServerID,
+					ServerNamespace: apidefaults.Namespace,
+				},
```
Should we omit `ServerMetadata` here since we know the access attempt was for an app, and we have app metadata below?

zmb3 reviewed (commented):
The PR title could be more descriptive. It refers to "session start" events, but the change is actually about preventing "app session start" events from being emitted for sessions that are unsuccessful.

21KennethTran commented on lib/auth/sessions.go:
```diff
@@ -544,6 +539,55 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
+			deviceTrustRequired = true
+			break
+		}
+	}
```
That is true. I am not sure if we really need all of the other checks in `roleset.checkAccess()` like label matching or login usernames, before issuing an app session certificate, which from my understanding comes from the Auth server. The access checks comes after the proxy connects to the app service, which should be when authorization takes place.

Let me know if doing the access checks earlier would be the better choice here. I could also refactor the device trust section into a utility function to use in these two parts.

codingllama commented on lib/auth/sessions.go:
```diff
@@ -537,6 +539,55 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
+		}
+	}
+
+	// Validate that the request has the deviceID extension; non-empty value means user authenticated with that device with CA
```
> authenticated with that device with CA

What does "CA" mean? Is this sentence complete?

21KennethTran replied:
I meant to refer to Teleport's Certificate Authority. I changed the comment for more clarity.

codingllama commented on lib/auth/sessions_test.go:
```diff
@@ -159,3 +163,79 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+	}
+
+	require.False(t, foundSuccess, "BUG: 'app.session.start' was emitted.")
+	require.True(t, foundFailure, "Missing 'auth.fail' (AuthAttempt) event for device trust rejection.")
```
```suggestion
	assert.False(t, foundSuccess, "BUG: 'app.session.start' was emitted.")
	assert.True(t, foundFailure, "Missing 'auth.fail' (AuthAttempt) event for device trust rejection.")
```

Also note that you don't really need the booleans, you could t.Errorf() directly where you set the booleans to true.

codingllama commented on lib/auth/sessions.go:
```diff
@@ -554,6 +555,45 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
+			},
+			Status: apievents.Status{
+				Success:     false,
+				Error:       errMsg,
```
Maybe, so we capture the original error too?

```suggestion
				Error:       err.Error(),
```

21KennethTran commented on lib/auth/sessions.go:
```diff
@@ -554,6 +555,45 @@ func (a *Server) CreateAppSessionFromReq(ctx context.Context, req NewAppSessionR
+	if err = checker.CheckDeviceAccess(services.AccessState{
+		DeviceVerified:           req.DeviceExtensions.DeviceID != "",
+		EnableDeviceVerification: true,
+		IsBot:                    req.Identity.IsBot(),
```
Nice catch. It seems that the `NewAppSessionRequest` has an `Identity: tlsca.Identity` field that never gets assigned to anything. I also noticed that the `isBot` and `BotInstanceID` fields are not being set in `CreateAppSession` even though they are called on in `CreateAppSessionFromReq`, so I added them to stay consistent with these changes. Let me know if this could cause any other problems down the line.

21KennethTran commented on lib/auth/sessions_test.go:
```diff
@@ -196,3 +200,80 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+	_, err = authServer.CreateUser(ctx, user)
+	require.NoError(t, err)
+
+	_, err = authServer.CreateAppSessionFromReq(ctx, auth.NewAppSessionRequest{
```
I changed the test case into a table driven one. Let me know if there are some other cases you would like to see!

codingllama commented on lib/auth/sessions_test.go:
```diff
@@ -211,69 +211,116 @@ func TestCreateAppSession_DeviceTrust(t *testing.T) {
+			trustMode:   "",
+			deviceID:    "",
+			expectError: false,
 		},
```
Add a success test using mode "required"?

codingllama commented on lib/auth/sessions_test.go:
```diff
@@ -196,3 +200,127 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+			logEntries, _, err := testAuthServer.AuditLog.SearchEvents(baseCtx, events.SearchEventsRequest{
+				From:  time.Now().Add(-1 * time.Minute),
+				To:    time.Now().Add(1 * time.Minute),
+				Limit: 20,
```
```suggestion
				Limit: 100, // arbitrary, enough to allow for events from concurrent tests
```

codingllama commented:
Please add a changelog entry to the end of the PR description.

codingllama commented on lib/auth/sessions_test.go:
```diff
@@ -196,3 +202,176 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+					Key:   backend.NewKey("devices", tt.deviceExtensions.DeviceID),
+					Value: deviceData,
+				})
+				require.NoError(t, err)
```
All of this is heavily made up, this isn't what a stored device looks like. That isn't the backend key for a device either. Plus, the storage backend for devices isn't plugged in here.

Do we actually need any of this?

tigrato commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -2948,6 +2948,13 @@ message AppSessionStart {
     (gogoproto.jsontag) = ""
   ];
+
+  // Status indicates the outcome of the event.
```
I agree with you.

There is no need to include success since the code already tells us that. We only care about adding the error and the user message which are fully compatible with the previous audit logs.

If we later decide that having success makes sense, we can always derive it from the error code before returning the event from the auth server

21KennethTran commented:
Thank you again for the comments. I spent a few days to fully flesh out these changes and test them locally. I have updated the After Demo in my original PR comment with the changes in the UI and more detailed audits. The test cases now test `CreateAppSession` instead of `CreateAppSessionFromReq` specific to device trust, and we also fixed the situation where previously `BotID` and `BotInstanceID` were not being passed correctly to `CreateAppSessionFromReq`. As always, another round of reviews would be appreciated!

codingllama commented on lib/auth/sessions_test.go:
```diff
@@ -196,3 +206,232 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+					SessionMetadata: apievents.SessionMetadata{
+						PrivateKeyPolicy: "",
+						WithMFA:          "",
+					},
```
```suggestion
```

codingllama reviewed (commented):
Looks good, just a handful of small suggestions.

21KennethTran commented on lib/auth/sessions_test.go:
```diff
@@ -196,3 +206,232 @@ func TestServer_CreateWebSessionFromReq_deviceWebToken(t *testing.T) {
+						apievents.ConnectionMetadata{},
+						apievents.ServerMetadata{},
+						apievents.DeviceMetadata{},
+					),
```
Yeah, it causes some problems when I use `protocmp.IgnoreFields` to ignore some generated fields in the metadata.

```
apievents.Metadata does not implement protoreflect.ProtoMessage (missing method ProtoReflect)
```
